pymongo/mongo_client.py

```diff
@@ -1602,7 +1603,12 @@ def _process_periodic_tasks(self):
 
     def __start_session(self, implicit, **kwargs):
         # Raises ConfigurationError if sessions are not supported.
-        server_session = self._get_server_session()
+        if implicit:
+            with self._topology._lock:
+                self._topology._check_session_support()
```

Comment:
We should never take the topology lock outside of methods in the Topology class. Let's make this a new topology method or maybe just a flag to the existing get_server_session() method. EG:
```
server_session = self._get_server_session(implicit)
```